### ISSUE TITLE
Enhancements to the Event class

### ISSUE THREAD

**Issue by Savageman** [fuel/core#829]
This issue is open to discussions.
- Allow arguments to be passed by references ;
- Add a priority system.

If you are interested in this, I'll probably make pull requests.

**Comment by WanWizard** [comment]
I find passing by reference potentially dangerous. 

It is designed to allow you to pass a callback to something that has access to the data that needs to be referenced or manipulated by the event method.

As for a priority system, I'm looking forward to a suggestion that doesn't break backward compatibility.

**Comment by Savageman** [comment]
For now, I'm using the Event class like this to handle references: 

```
Event::register('event_name', function() {
    return function(&$ref) { }
});
foreach (Event::trigger('event_name', null ,'array') as $callback) {
    is_callable($c) and call_user_func_array($c, array(&$ref_arg));
}
```

About priorities, we could just add a 3rd parameter to the `register()` method with a default value `$priority = 10` (default value used by Wordpress).
Pretty straigtforward but effective and backwards compatible.

**Comment by WanWizard** [comment]
Could you send a pull request with your suggestion for implementation?

I won't be having time for this any time soon.

**Comment by Savageman** [comment]
Not a pull request, but something you can look at:

https://github.com/Savageman/core/commit/d3065dd50ff50a93ab72ca3e6544ee36f577f8de (reminder to self: I forgot `ksort()`)

Usage:

```
<?php
// 7 is the priority
Event::register('name', 7, 'callback');
```

It's not perfect because the priority is set depending on the type of the 2nd argument (int or callable), but needed for backwards compatibility.
Because all remaining arguments after the callback are passed to the callback itself, I cannot add another parameter at the end.

**Comment by jschreuder** [comment]
This is more a 2.0 change, I'm assigning this to the 2.0-dev milestone.

**Comment by Savageman** [comment]
Seems reasonable to me! :)

**Comment by huglester** [comment]
Anyone thought about those events, to be smth like 'mysql' driven? 
What I mean, I would like to 'queue' (insert) all those events to the database, 
and trigger them via crontab(by priority), or smth like that.

Or this job belongs to something else than Events ? 

Because I used Events to send emails - it is OK if you send emails using local mail server, 
but if for example you try to auth to Gmail servers - it could take up to 1-3 seconds...

**Comment by WanWizard** [comment]
That is more of a message queue implementation, some different then events, and something that more belongs in a package.

**Comment by frankdejonge** [comment]
This has been implemented in the 2.0 version.
